File: Applied Course/4.Problem Solving/7.Problems on Strings/8.Print all anagrams in a list of words.py
# https://leetcode.com/problems/group-anagrams/
# Logic-1
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams=collections.defaultdict(list)
        for word in strs:
            anagrams[tuple(sorted(word))].append(word)
        return [v for k,v in anagrams.items()]

# Logic-2
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagram=collections.defaultdict(list)
        for word in strs:
            count=[0]*26
            for c in word:
                count[ord(c)-ord('a')]+=1
            anagram[tuple(count)].append(word)
        return anagram.values()

# Logic-3
from collections import defaultdict 

def printAnagramsTogether(words):
    groupedWords = defaultdict(list) 
    
    for word in words:
        
        groupedWords["".join(sorted(word))].append(word) 
    
    for group in groupedWords.values(): 
        print(" ".join(group))


if __name__ == "__main__": 
    arr = ["cat", "dog", "tac", "god", "act"] 
    printAnagramsTogether(arr)

# Logic-4
arr = ['cat', 'tac', 'man', 'dog', 'nam', 'bbt']
H = {}
for word in arr:
    d = [0]*26
    for letter in word:
        ascii_val = ord(letter)
        d[ascii_val - ord('a')]+=1
    word_key = tuple(d)
    if word_key in H:
        H[word_key].append(word)
    else:
        H[word_key] = [word]

print("All anagrams are: ")
for k, v in H.items():
    if len(v) > 1:
        for word in v:
            print(word, end=' ')

File: Applied Course/4.Problem Solving/7.Problems on Strings/8.Print all anagrams in a list of words.cpp
#include<bits/stdc++.h>
using namespace std;
// Logic-1
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string,vector<string>> mymap;
        int n=strs.size();
        string temp;
        for(int i=0;i<n;i++)
        {
            temp=strs[i];
            sort(strs[i].begin(),strs[i].end());
            mymap[strs[i]].push_back(temp);
        }
        vector<vector<string>> result;
        for(auto itr=mymap.begin();itr!=mymap.end();itr++)
            result.push_back(itr->second);
        return result;
    
};
int main()
{
    // https://www.youtube.com/watch?v=0I6IL3TnIZs
    // https://leetcode.com/problems/group-anagrams
    return 0;
}